Aplib.Core/Belief/MemoryBelief.cs

```diff
@@ -0,0 +1,104 @@
+using System;
+
+namespace Aplib.Core.Belief
+{
+    /// <summary>
+    /// The <see cref="MemoryBelief{TReference, TObservation}"/> class represents a belief with "memory".
+    /// Some <i>object reference</i> is used to generate/update a <i>observation</i> 
+    /// (i.e., some piece of information on the game state as perceived by an agent).
+    /// This belief also stores a limited amount of previous observations in memory.
+    /// </summary>
+    /// <remarks>
+    /// It implements the <see cref="IBelief"/> interface.
+    /// It supports implicit conversion to <typeparamref name="TObservation"/>.
+    /// </remarks>
+    /// <typeparam name="TReference">The type of the reference used to generate/update the resource.</typeparam>
+    /// <typeparam name="TObservation">The type of the resource the belief represents.</typeparam>
+    public class MemoryBelief<TReference, TObservation> : Belief<TReference, TObservation>
+    {
+        /// <summary>
+        /// A "memorized" resouce, from the last time the belief was updated.
+        /// </summary>
+        private readonly CircularArray<TObservation> _memorizedObservations;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryBelief{TReference, TResource}"/> class.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the resource.</param>
+        /// <param name="getResourceFromReference">A function that takes a reference and generates/updates a resource.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        public MemoryBelief(TReference reference, Func<TReference, TObservation> getResourceFromReference, int framesToRemember)
+            : base(reference, getResourceFromReference, () => true)
```

Comment:
Can remove the lambda.
```suggestion
        public MemoryBelief(TReference reference, Func<TReference, TObservation> getObservationFromReference, int framesToRemember)
            : base(reference, getObservationFromReference)
```